File: frontend/src/types/participants.ts
import type { MonitoredAccount } from "./accounts";

export interface Snapshot {
  cpa_contract_known?: boolean;
  participant_id: number;
  participant_name: string;
  source_sub2api_user_id?: number | null;
  quota_pool_id?: number | null;
  quota_pool_name?: string;
  pool_contract_revision?: number | null;
  share_percent?: number;
  selected_cost: number;
  delta_cost: number | null;
  charged_delta_percent: number;
  charged_cycle_percent: number;
  charged_percent_lower: number | null;
  charged_percent_upper: number | null;
  remaining_share_percent: number | null;
  current_balance_usd: number | null;
  recommended_balance_usd: number | null;
  recommended_balance_min_usd: number | null;
  recommended_balance_max_usd: number | null;
  deterministic_balance_min_usd: number | null;
  deterministic_balance_max_usd: number | null;
  balance_difference_usd: number | null;
  is_overused: boolean;
  overused_percent: number;
  overused_percent_min: number;
  overused_percent_max: number;
  needs_manual_update: boolean;
  recommendation_applied: boolean;
  reason: string;
  allocation_model: "time_varying" | "constant_average";
}
export interface AccountBreakdown {
  id: number | null;
  account_id: number;
  external_account_id: number;
  account_name: string;
  account_enabled: boolean;
  pool_id: number;
  pool_name: string;
  contract_share_percent: number;
  allocated: boolean;
  latest_selected_cost: number | null;
  last_checked_at: string | null;
  snapshot: Snapshot | null;
}
export interface AggregateRecommendationSource {
  account_id: number;
  external_account_id: number;
  account_name: string;
  pool_id: number;
  pool_name: string;
  pool_contract_revision: number;
  contract_share_percent: number;
  snapshot: Snapshot | null;
  net_position_usd: number | null;
  net_position_min_usd: number | null;
  net_position_max_usd: number | null;
  contribution_usd: number | null;
  contribution_min_usd: number | null;
  contribution_max_usd: number | null;
  estimated_capacity_usd: number | null;
  expected_entitlement_usd: number | null;
  consumed_entitlement_usd: number | null;
  remaining_entitlement_usd: number | null;
  entitlement_usage_percent: number | null;
}
export interface ParticipantPoolAllocation {
  pool_id: number;
  pool_name: string;
  share_percent: number;
  account_ids?: number[];
  account_count?: number;
}
export interface AggregateRecommendation {
  participant_id: number;
  participant_name: string;
  pool_allocations: ParticipantPoolAllocation[];
  selected_cost: number;
  charged_cycle_percent: number;
  expected_entitlement_usd: number | null;
  consumed_entitlement_usd: number | null;
  remaining_entitlement_usd: number | null;
  entitlement_usage_percent: number | null;
  current_balance_usd: number | null;
  recommended_balance_usd: number | null;
  recommended_balance_min_usd: number | null;
  recommended_balance_max_usd: number | null;
  balance_difference_usd: number | null;
  is_overused: boolean;
  needs_manual_update: boolean;
  recommendation_applied: boolean;
  recommendation_complete: boolean;
  account_count: number;
  pool_count: number;
  reason: string;
  allocation_model: "partitioned_pool_sum";
  sources: AggregateRecommendationSource[];
}
export interface Participant {
  id: number;
  name: string;
  email: string;
  sub2api_user_id: number | null;
  sub2api_username: string;
  sub2api_email: string;
  sub2api_identity: string;
  pool_allocations: ParticipantPoolAllocation[];
  is_owner: boolean;
  enabled: boolean;
  notes: string;
  latest_balance_usd: number | null;
  last_checked_at: string | null;
  account_breakdowns: AccountBreakdown[];
  snapshot: AggregateRecommendation | null;
}
export interface QuotaAllocationParticipant {
  id: number;
  name: string;
  sub2api_user_id: number | null;
  sub2api_username: string;
  sub2api_email: string;
  sub2api_identity: string;
  is_owner: boolean;
  enabled: boolean;
}
export interface QuotaPoolAllocationEntry {
  participant_id: number;
  share_percent: number;
}
export interface QuotaPoolAllocation {
  id: number;
  name: string;
  contract_revision: number;
  account_ids: number[];
  allocations: QuotaPoolAllocationEntry[];
  total_share_percent: number;
}
export interface QuotaAllocationData {
  provider?: "sub2api" | "cpa" | "gpt_load";
  accounts: MonitoredAccount[];
  participants: QuotaAllocationParticipant[];
  pools: QuotaPoolAllocation[];
}
export interface QuotaAllocationWritePool {
  id?: number;
  name: string;
  account_ids: number[];
  allocations: QuotaPoolAllocationEntry[];
}
export interface QuotaAllocationWrite {
  provider?: "sub2api" | "cpa" | "gpt_load";
  pools: QuotaAllocationWritePool[];
}
export interface Sub2APIUserOption {
  id: number;
  email: string;
  username: string;
  status: string;
  role: string;
}

File: backend/monitor/openapi/participants.py
"""OpenAPI schemas for participants and aggregate recommendations."""

from __future__ import annotations


def participant_schemas(nullable_number: dict, nullable_string: dict) -> dict:
    return {
        "ParticipantPoolAllocation": {
            "type": "object",
            "required": [
                "pool_id",
                "pool_name",
                "share_percent",
                "account_ids",
            ],
            "properties": {
                "pool_id": {"type": "integer"},
                "pool_name": {"type": "string"},
                "share_percent": {"type": "number"},
                "account_ids": {
                    "type": "array",
                    "items": {"type": "integer"},
                },
            },
        },
        "Participant": {
            "type": "object",
            "required": [
                "id",
                "name",
                "email",
                "sub2api_user_id",
                "sub2api_username",
                "sub2api_email",
                "sub2api_identity",
                "pool_allocations",
                "is_owner",
                "enabled",
                "notes",
                "latest_balance_usd",
                "last_checked_at",
                "account_breakdowns",
                "snapshot",
            ],
            "properties": {
                "id": {"type": "integer"},
                "name": {"type": "string"},
                "email": {"type": "string"},
                "sub2api_user_id": {"type": ["integer", "null"]},
                "sub2api_username": {"type": "string"},
                "sub2api_email": {"type": "string"},
                "sub2api_identity": {"type": "string"},
                "pool_allocations": {
                    "type": "array",
                    "items": {
                        "$ref": "#/components/schemas/ParticipantPoolAllocation"
                    },
                },
                "is_owner": {"type": "boolean"},
                "enabled": {"type": "boolean"},
                "notes": {"type": "string"},
                "latest_balance_usd": nullable_number,
                "last_checked_at": nullable_string,
                "account_breakdowns": {
                    "type": "array",
                    "items": {
                        "$ref": "#/components/schemas/AccountBreakdown"
                    },
                },
                "snapshot": {
                    "oneOf": [
                        {
                            "$ref": "#/components/schemas/AggregateRecommendation"
                        },
                        {"type": "null"},
                    ]
                },
            },
        },
        "ParticipantSnapshot": {
            "type": "object",
            "description": (
                "selected_cost 保持修正后的测算成本口径；建议余额、余额范围与差额"
                "已按当前区间历史消费结构换算为预计 Sub2API 实际扣费金额。"
                "余额范围以该估计倍率为条件，不保证未来请求构成不变。"
            ),
            "required": [
                "cpa_contract_known",
                "participant_id",
                "participant_name",
                "quota_pool_id",
                "quota_pool_name",
                "pool_contract_revision",
                "share_percent",
                "selected_cost",
                "delta_cost",
                "charged_delta_percent",
                "charged_cycle_percent",
                "charged_percent_lower",
                "charged_percent_upper",
                "remaining_share_percent",
                "current_balance_usd",
                "recommended_balance_usd",
                "recommended_balance_min_usd",
                "recommended_balance_max_usd",
                "deterministic_balance_min_usd",
                "deterministic_balance_max_usd",
                "balance_difference_usd",
                "is_overused",
                "overused_percent",
                "overused_percent_min",
                "overused_percent_max",
                "needs_manual_update",
                "recommendation_applied",
                "reason",
                "allocation_model",
            ],
            "properties": {
                "cpa_contract_known": {"type": "boolean", "description": "CPA 观测是否存在当时生效的份额依据；无依据时剩余份额未知。"},
                "participant_id": {"type": "integer"},
                "participant_name": {"type": "string"},
                "quota_pool_id": {
                    "type": ["integer", "null"]
                },
                "quota_pool_name": {"type": "string"},
                "pool_contract_revision": {
                    "type": ["integer", "null"]
                },
                "share_percent": {"type": "number"},
                "selected_cost": {"type": "number"},
                "delta_cost": nullable_number,
                "charged_delta_percent": {"type": "number"},
                "charged_cycle_percent": {"type": "number"},
                "charged_percent_lower": nullable_number,
                "charged_percent_upper": nullable_number,
                "remaining_share_percent": nullable_number,
                "current_balance_usd": nullable_number,
                "recommended_balance_usd": nullable_number,
                "recommended_balance_min_usd": nullable_number,
                "recommended_balance_max_usd": nullable_number,
                "deterministic_balance_min_usd": nullable_number,
                "deterministic_balance_max_usd": nullable_number,
                "balance_difference_usd": nullable_number,
                "is_overused": {"type": "boolean"},
                "overused_percent": {"type": "number"},
                "overused_percent_min": {"type": "number"},
                "overused_percent_max": {"type": "number"},
                "needs_manual_update": {"type": "boolean"},
                "recommendation_applied": {"type": "boolean"},
                "reason": {"type": "string"},
                "allocation_model": {
                    "type": "string",
                    "enum": ["time_varying", "constant_average"],
                },
            },
        },
        "AccountBreakdown": {
            "type": "object",
            "required": [
                "id",
                "account_id",
                "external_account_id",
                "account_name",
                "account_enabled",
                "pool_id",
                "pool_name",
                "contract_share_percent",
                "allocated",
                "snapshot",
                "latest_selected_cost",
                "last_checked_at",
            ],
            "properties": {
                "id": {
                    "oneOf": [
                        {"type": "integer"},
                        {"type": "null"},
                    ]
                },
                "account_id": {"type": "integer"},
                "external_account_id": {"type": "integer"},
                "account_name": {"type": "string"},
                "account_enabled": {"type": "boolean"},
                "pool_id": {"type": "integer"},
                "pool_name": {"type": "string"},
                "contract_share_percent": {"type": "number"},
                "allocated": {"type": "boolean"},
                "latest_selected_cost": nullable_number,
                "last_checked_at": nullable_string,
                "snapshot": {
                    "oneOf": [
                        {
                            "$ref": "#/components/schemas/ParticipantSnapshot"
                        },
                        {"type": "null"},
                    ]
                },
            },
        },
        "AggregateRecommendationSource": {
            "description": (
                "net_position 与 contribution 金额为已换算的 Sub2API 钱包口径，"
                "各来源带符号换算后再汇总；capacity 和 entitlement 指标仍为"
                "修正后的测算权益口径。"
            ),
            "type": "object",
            "required": [
                "account_id",
                "external_account_id",
                "account_name",
                "pool_id",
                "pool_name",
                "pool_contract_revision",
                "contract_share_percent",
                "snapshot",
                "net_position_usd",
                "net_position_min_usd",
                "net_position_max_usd",
                "contribution_usd",
                "contribution_min_usd",
                "contribution_max_usd",
                "estimated_capacity_usd",
                "expected_entitlement_usd",
                "consumed_entitlement_usd",
                "remaining_entitlement_usd",
                "entitlement_usage_percent",
            ],
            "properties": {
                "account_id": {"type": "integer"},
                "external_account_id": {"type": "integer"},
                "account_name": {"type": "string"},
                "pool_id": {"type": "integer"},
                "pool_name": {"type": "string"},
                "pool_contract_revision": {"type": "integer"},
                "contract_share_percent": {"type": "number"},
                "carry_adjustment_percent": {"type": "number", "description": "本周期借用结转调整，单位为百分点，不改写合同份额。"},
                "effective_share_percent": {"type": "number", "description": "合同份额加本周期结转后的可用权益。"},
                "net_position_usd": nullable_number,
                "net_position_min_usd": nullable_number,
                "net_position_max_usd": nullable_number,
                "estimated_capacity_usd": nullable_number,
                "expected_entitlement_usd": nullable_number,
                "consumed_entitlement_usd": nullable_number,
                "remaining_entitlement_usd": nullable_number,
                "entitlement_usage_percent": nullable_number,
                "snapshot": {
                    "oneOf": [
                        {
                            "$ref": "#/components/schemas/ParticipantSnapshot"
                        },
                        {"type": "null"},
                    ]
                },
                "contribution_usd": nullable_number,
                "contribution_min_usd": nullable_number,
                "contribution_max_usd": nullable_number,
            },
        },
        "AggregateRecommendation": {
            "type": "object",
            "description": (
                "recommended_balance 及其范围、差额是预计应设置的 Sub2API 全局余额，"
                "不是追加充值金额；展示、通知及写入方不得再次应用计费修正倍率。"
                "临时爽蹬生效时完整建议固定为 9999，退出后恢复普通建议并计入周期权益调整。"
            ),
            "required": [
                "participant_id",
                "participant_name",
                "pool_allocations",
                "selected_cost",
                "charged_cycle_percent",
                "expected_entitlement_usd",
                "consumed_entitlement_usd",
                "remaining_entitlement_usd",
                "entitlement_usage_percent",
                "current_balance_usd",
                "recommended_balance_usd",
                "recommended_balance_min_usd",
                "recommended_balance_max_usd",
                "balance_difference_usd",
                "is_overused",
                "needs_manual_update",
                "recommendation_applied",
                "recommendation_complete",
                "account_count",
                "pool_count",
                "reason",
                "allocation_model",
                "sources",
            ],
            "properties": {
                "temporary_burst": {"type": "boolean", "description": "当前是否处于临时统一余额模式。"},
                "temporary_burst_expires_at": nullable_string,
                "participant_id": {"type": "integer"},
                "participant_name": {"type": "string"},
                "pool_allocations": {
                    "type": "array",
                    "items": {
                        "$ref": "#/components/schemas/ParticipantPoolAllocation"
                    },
                },
                "selected_cost": {"type": "number"},
                "charged_cycle_percent": {"type": "number"},
                "expected_entitlement_usd": nullable_number,
                "consumed_entitlement_usd": nullable_number,
                "remaining_entitlement_usd": nullable_number,
                "entitlement_usage_percent": nullable_number,
                "current_balance_usd": nullable_number,
                "recommended_balance_usd": nullable_number,
                "recommended_balance_min_usd": nullable_number,
                "recommended_balance_max_usd": nullable_number,
                "balance_difference_usd": nullable_number,
                "is_overused": {"type": "boolean"},
                "needs_manual_update": {"type": "boolean"},
                "recommendation_applied": {"type": "boolean"},
                "recommendation_complete": {"type": "boolean"},
                "account_count": {"type": "integer"},
                "pool_count": {"type": "integer"},
                "reason": {"type": "string"},
                "allocation_model": {
                    "type": "string",
                    "const": "partitioned_pool_sum",
                },
                "sources": {
                    "type": "array",
                    "items": {
                        "$ref": "#/components/schemas/AggregateRecommendationSource"
                    },
                },
            },
        },
        "AppliedRecommendation": {
            "type": "object",
            "required": [
                "operation_id",
                "participant_id",
                "sub2api_user_id",
                "applied_balance_usd",
                "account_count",
            ],
            "properties": {
                "operation_id": {"type": "string", "format": "uuid"},
                "participant_id": {"type": "integer"},
                "sub2api_user_id": {"type": ["integer", "null"]},
                "applied_balance_usd": {"type": "number"},
                "account_count": {"type": "integer"},
            },
        },
    }
